[DR-3123] Migrate to vite

samanehsan opened the pull request:
This PR updates the UI to use Vite instead of create react app (`react-scripts` is one of the dependencies, which has not been maintained for over a year and contains a security vulnerability: https://broadworkbench.atlassian.net/browse/DR-3123).

Some things to note:
- Vite has a built-in hot reloader
- The route proxying is defined in vite.config.js (so setupProxy is no longer needed)
- When running locally, the url needs to use "localhost" instead of "127.0.0.1" in order for log in to work

I also updated Cypress from version 9 -> 12, which supports the latest version of Vite. This involved:
- Replacing `cypress.json` with `cypress.config.ts`
- Removing the `cypress/plugins/index.js` file
- Adding a `cypress/support/component-index.html` file (where the component test html gets rendered)
- Updating the syntax in the e2e tests (e.g. cy.route is now cy.intercept) and some of the unit/component tests

samanehsan committed:
Install vite related libraries and uninstall react-scripts

samanehsan committed:
Update package.json scripts

samanehsan committed:
Create vite config

samanehsan committed:
Fix index.html

samanehsan committed:
Get UI to run locally

samanehsan committed:
Fix eslint

samanehsan committed:
Fix cypress tests

samanehsan committed:
Upgrade to Cypress 12 and fix component tests

samanehsan committed:
Fix cypress e2e tests

samanehsan committed:
Fix middleware

samanehsan committed:
Remove commented code

samanehsan commented:
The test datasets/snapshots used by the e2e tests exist but for some reason are not accessible to the test user (I can't access them with my dev account either). I created new ones in int-6 as a temporary work-around so I could run/fix the e2e tests locally, so they should pass once we can sort out the issue with int-4 and int-5!

samanehsan committed:
Ignore sonarcloud errors

- The gcloud libraries are not pinned to a version so
  we cannot use an integrity hash here
- Fontawesome no longer supports SRI hashes

samanehsan commented:
The SonarCloud failure can be resolved by generating a hash for our fontawesome kit "https://kit.fontawesome.com/316fe749fb.js", but I get a 403 error when I try to do that (I used this website: https://www.srihash.org/). Does anyone have access? Or should I create a new custom kit?

okotsopoulos commented:
@samanehsan that FontAwesome question rings a bell, I'm not sure if this is possible :(

From @snf2ye in February '23:
> Fontawesome supports the use an SRI hash with versions <5. We are using a fontawesome kit v6. Fontawesome made a product choice to no longer support SRI hashes ([source](https://github.com/FortAwesome/Font-Awesome/issues/16691)).

https://broadworkbench.atlassian.net/browse/DR-2872

samanehsan commented:
Thanks @okotsopoulos! That's helpful context! I'll mark that script as excluded

samanehsan committed:
Clean up unused dependencies

samanehsan commented:
I think the e2e tests aren't running because it waits on `localhost:3000` but the vite dev server is using `127.0.0.1:3000`. I'll see if there's a way to change the server to use localhost.

s-rubenstein commented on src/components/common/TextContent.test.tsx:
```diff
@@ -13,7 +13,7 @@ const markdownText = '# Lorem Ipsum \n\n[Link](https://broadinstitute.org)';
 const htmlText = '<a href="www.badsite.com">A bad day</a>';
 
-describe('WithoutStylesMarkdownContent', () => {
+describe('TextContent', () => {
```
Why the name change here? Is it sonar related? This seems less descriptive.

samanehsan replied:
`WithoutStylesMarkdownContent` is the name of a different component, I thought it was a copy-paste error

s-rubenstein commented on src/components/common/WithoutStylesMarkdownContent.test.tsx:
```diff
@@ -39,9 +39,9 @@ describe('WithoutStylesMarkdownContent', () => {
-    cy.get('div#__cy_root > div > p').should('contain.text', 'A bad day');
-    cy.get('div#__cy_root > div > p').children().should('have.length', 0);
+    cy.root().should('have.length', 1);
+    cy.root().get('p').should('contain.text', 'A bad day');
```
Just as a note, do you think this sort of test could maybe be simplified by doing a findByText on the page? That is what Terra-UI does

samanehsan replied:
Ah oops I missed this comment! It turns out `findByText` requires another dependency so I'm inclined to leave it out for now if that's okay

s-rubenstein commented on src/constants/index.ts:
```diff
@@ -1,5 +1,4 @@
 import { CloudPlatform } from 'generated/tdr';
-import _ from 'lodash';
```
I'm surprised you can remove the import from lodash without any other changes to the file. Do you have any linting for unused imports?

samanehsan replied:
Good point! Our eslint rules don't include unused imports so I created a ticket for that: https://broadworkbench.atlassian.net/browse/DR-3176

s-rubenstein commented on src/reducers/snapshot.ts:
```diff
@@ -352,13 +352,13 @@ export default {
           return immutable(state, {
-            lastPath: { $set: path },
-          });
+            lastPath: { $set: path as string },
```
I'm surprised you have to do the casts here. Does typescript not type interpolate correctly for these?

samanehsan replied:
Yeah, weirdly there were a couple of places where the type couldn't be inferred and without making it explicit the server wouldn't run

s-rubenstein replied:
Ah, in this particular case it might be that it thinks path is `string | undefined` given the ?s. Would you file a ticket to look into the specific casts? My guess is that they are resolvable using typescript but there may be some code structure changes needed.

samanehsan replied:
Here's the ticket! https://broadworkbench.atlassian.net/browse/DR-3177

s-rubenstein reviewed (commented):
I'm noticing a lot more casting than was there before - is this a result of the libraries changing? Casting like that can sometimes be a code smell, so I am curious if you know where its coming from.

samanehsan committed:
Use localhost instead of 127.0.0.1

Also reenable linting on server start

samanehsan committed:
Fix e2e test wait-on local server

Work around wait-on 404 error

Originally from: https://github.com/vitejs/vite/issues/9520#issuecomment-1660050942

samanehsan committed:
Try replacing wait-on

samanehsan committed:
Update to from node 14->16

Fixes build issue with vite

samanehsan commented on src/store/middleware.ts:
```diff
@@ -7,7 +7,7 @@ export const sagaMiddleware = createSagaMiddleware();
 
 /* istanbul ignore next */
-if (process.env.NODE_ENV === 'development') {
+if (import.meta.env.DEV) {
```
This env variable is provided by Vite by default: https://vitejs.dev/guide/env-and-mode.html#env-variables

samanehsan committed:
Remove unnecessary --production flag

vite builds production by default

samanehsan committed:
Use default vite react plugin config

samanehsan committed:
Try defining process

samanehsan committed:
Try updating jsondiffpatch version

samanehsan commented:
The "process not defined" issue causing the e2e tests to fail was due to to the version of jsondiffpatch we were using. Updating to the latest version fixed the issue! 🎉 

- https://github.com/reduxjs/redux-devtools/issues/937#issuecomment-962520662
- https://github.com/benjamine/jsondiffpatch/issues/315

samanehsan committed:
Remove unused packages

samanehsan committed:
Update .node-version

samanehsan committed:
Fix cypress.config.ts

s-rubenstein commented on Dockerfile.direct:
```diff
@@ -10,15 +10,15 @@ ARG TDR_OPEN_API_YAML_LOCATION
 # Run the build
 RUN export DISABLE_ESLINT_PLUGIN=true && \
   npm ci && \
-  npm run build-no-code-gen --production
```
Is this flag removed?

samanehsan replied:
The `--production` flag isn't needed because `vite build` uses production by default

s-rubenstein reviewed (approved):
LGTM

petesantos commented on package.json:
```diff
@@ -24,6 +24,10 @@
   "dependencies": {
     "@babel/runtime": "^7.21.5",
     "@codemirror/lang-javascript": "^6.1.5",
+    "@cypress/vite-dev-server": "^5.0.5",
```
I believe there are some deps that can be moved to `devDependencies`. This isn't a big deal, but more for `package.json` hygiene

Feel free to do this in a follow up PR

petesantos commented on package.json:
```diff
@@ -48,20 +51,19 @@
     "clsx": "^1.1.0",
     "connected-react-router": "^6.9.3",
-    "cypress": "^9.7.0",
+    "cypress": "^12.17.3",
```
This can be moved to `devDependencies`

petesantos commented on package.json:
```diff
@@ -97,25 +98,27 @@
     "tss-react": "^4.8.4",
     "typescript": "^5.0.4",
     "validator": "^13.9.0",
+    "vite": "^4.4.9",
```
I believe `vite` can be moved to `devDepenendcies`

samanehsan committed:
Package clean-up
